Fix distro fact handling for Flatcar

johananl opened the pull request:
##### SUMMARY
The existence of the file `/etc/flatcar/update.conf` depends on bootstrap configuration typically provided by the user. For that reason this file is unsuitable for determining distro facts for Flatcar Container Linux. This PR switches to `/etc/os-release` as the file used for extracting distro facts for Flatcar.

Fixes #77537.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

distribution.py

>NOTE: I'm having hard time figuring out how to test this change. There is `test/units/module_utils/facts/system/distribution/fixtures/flatcar_2492.0.0.json` but I couldn't get reliable results by executing the test which uses this fixture. Any help regarding how to test this change would be appreciated!

johananl committed:
Fix distro fact handling for Flatcar

The existence of the file /etc/flatcar/update.conf depends on
bootstrap configuration typically provided by the user. For that
reason this file is unsuitable for determining distro facts for
Flatcar Container Linux.

The distribution_release fact is meaningless in the case of Flatcar
since Flatcar doesn't have named releases. The distribution_version
fact, however, IS meaningful and should contain a number such as
"3139.2.0".

- Use /etc/os-release instead of /etc/flatcar/update.conf.
- Drop the distribution_release fact.
- Set the distribution_version fact.

Signed-off-by: Johanan Liebermann <jliebermann@microsoft.com>

pothos commented on lib/ansible/module_utils/facts/system/distribution.py:
```diff
@@ -447,15 +447,16 @@ def parse_distribution_file_Flatcar(self, name, data, path, collected_facts):
+        if not data:
+            return False, flatcar_facts
+
+        version = re.search("VERSION=(.*)", data)
```
Maybe we could even rely more on `distro` and reuse `_distro.version()` that already provides the `VERSION_ID` field, or?

johananl replied:
I like the idea, however `distro` inside `parse_distribution_file_Flatcar()` is a **string**, not an instance of `LinuxDistribution`. The `version()` method on `LinuxDistribution` is actually already used inside `get_distribution_version()` which gets invoked from `_guess_distribution()` which in turns runs before we try processing the dist file. The dist file parsing functions are **overriding** any facts guessed previously, and AFAICT are supposed to be the authoritative source for distro-specific facts.

Therefore, I think it's safer to explicitly set the relevant facts inside `parse_distribution_file_Flatcar()` and rely on guessed facts only as a fallback.

johananl committed:
Update distro test fixture for Flatcar

- Generate the fixture using gen_distribution_version_testcase.py.
- Override result.distribution and result.os_family manually as the
  generator script gives wrong values.
- Use a recent Flatcar version.

Signed-off-by: Johanan Liebermann <jliebermann@microsoft.com>

johananl commented:
>>NOTE: I'm having hard time figuring out how to test this change. There is `test/units/module_utils/facts/system/distribution/fixtures/flatcar_2492.0.0.json` but I couldn't get reliable results by executing the test which uses this fixture. Any help regarding how to test this change would be appreciated!

I've updated the existing test fixture. The test seems to pass, however I'm not totally sure about the `distro` key [here](https://github.com/ansible/ansible/pull/77635/files#diff-6821b161be4d93232945e9a11ac642ca6afeaa70c07690341df67ef82c27c073R7) and whether it is correct.

johananl commented:
> > The test seems to pass, however I'm not totally sure about the `distro` key [here](https://github.com/ansible/ansible/pull/77635/files#diff-6821b161be4d93232945e9a11ac642ca6afeaa70c07690341df67ef82c27c073R7) and whether it is correct.
> 
> That fixture can be generated by running `hacking/tests/gen_distribution_version_testcase.py`

Thanks @sivel. The generated output has `result.distribution` and `result.os_family` set to `Flatcar Container Linux by Kinvolk` though, which isn't what we want (see #77537) 😕
Generating the fixture and then setting these two keys to `Flatcar` manually seems to work well though.

s-hertel commented:
/azp run

webknjaz commented:
/azp run
